### PR TITLE
Add SocketException and 'insufficient data written' to retryable exceptions

### LINKED ISSUE #1180: Storage - SocketException Socket closed
Hi fellow Google cloud java users,
We are seeing Socket closed quite often. 
The StorageOptions.Builder().connectTimeout and StorageOptions.Builder().readTimeout are not set so I guess the default of 20 seconds is used.
I didn't want to change them as 20 seconds should be plenty. 

Does anyone have an idea about this?

```
java.net.SocketException: Socket closed
    at java.net.SocketInputStream.read
    at java.net.SocketInputStream.read
    at sun.security.ssl.InputRecord.readFully
    at sun.security.ssl.InputRecord.read
    at sun.security.ssl.SSLSocketImpl.readRecord
    at sun.security.ssl.SSLSocketImpl.readDataRecord
    at sun.security.ssl.AppInputStream.read
    at java.io.BufferedInputStream.fill
    at java.io.BufferedInputStream.read1
    at java.io.BufferedInputStream.read
    at sun.net.www.http.HttpClient.parseHTTPHeader
    at sun.net.www.http.HttpClient.parseHTTP
    at sun.net.www.protocol.http.HttpURLConnection.getInputStream0
    at sun.net.www.protocol.http.HttpURLConnection.getInputStream
    at java.net.HttpURLConnection.getResponseCode
    at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode
    at com.google.api.client.http.javanet.NetHttpResponse.<init>
    at com.google.api.client.http.javanet.NetHttpRequest.execute
    at com.google.api.client.http.HttpRequest.execute
    at com.google.cloud.storage.spi.DefaultStorageRpc.write
    at com.google.cloud.storage.BlobWriteChannel$1.run
    at java.util.concurrent.Executors$RunnableAdapter.call
    at com.google.cloud.RetryHelper.doRetry
    at com.google.cloud.RetryHelper.runWithRetries
    at com.google.cloud.RetryHelper.runWithRetries
    at com.google.cloud.storage.BlobWriteChannel.flushBuffer
    at com.google.cloud.BaseWriteChannel.close
    at com.spotify.buildartifactarchiver.storage.GcsUploaderClient.storeInputStream
    at com.spotify.buildartifactarchiver.storage.ArtifactUploader.uploadData
    at com.spotify.buildartifactarchiver.ArtifactArchiver.archiveRequest
    at com.spotify.buildartifactarchiver.ArtifactArchiver.archiveBuild
    at com.spotify.buildartifactarchiver.web.ArtifactArchiverPubSubResource.messageHandler
    at com.spotify.google.cloud.pubsub.client.Puller.lambda$pullBatch$11
    at java.util.concurrent.CompletableFuture.uniWhenComplete
    at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire
    at java.util.concurrent.CompletableFuture.postComplete
    at java.util.concurrent.CompletableFuture.complete
    at com.spotify.google.cloud.pubsub.client.PubsubFuture.succeed
    at com.spotify.google.cloud.pubsub.client.PubsubFuture.lambda$wrap$9
    at java.util.concurrent.CompletableFuture.uniWhenComplete
    at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire
    at java.util.concurrent.CompletableFuture.postComplete
    at java.util.concurrent.CompletableFuture.complete
    at com.spotify.google.cloud.pubsub.client.PubsubFuture.succeed
    at com.spotify.google.cloud.pubsub.client.Pubsub.lambda$requestJavaNet$8
    at java.util.concurrent.ThreadPoolExecutor.runWorker
    at java.util.concurrent.ThreadPoolExecutor$Worker.run
    at java.lang.Thread.run
com.google.cloud.storage.StorageException: Socket closed
    at com.google.cloud.storage.spi.DefaultStorageRpc.translate
    at com.google.cloud.storage.spi.DefaultStorageRpc.write
    at com.google.cloud.storage.BlobWriteChannel$1.run
    at java.util.concurrent.Executors$RunnableAdapter.call
    at com.google.cloud.RetryHelper.doRetry
    at com.google.cloud.RetryHelper.runWithRetries
    at com.google.cloud.RetryHelper.runWithRetries
    at com.google.cloud.storage.BlobWriteChannel.flushBuffer
    at com.google.cloud.BaseWriteChannel.close
    at com.spotify.buildartifactarchiver.storage.GcsUploaderClient.storeInputStream
    at com.spotify.buildartifactarchiver.storage.ArtifactUploader.uploadData
    at com.spotify.buildartifactarchiver.ArtifactArchiver.archiveRequest
    at com.spotify.buildartifactarchiver.ArtifactArchiver.archiveBuild
    at com.spotify.buildartifactarchiver.web.ArtifactArchiverPubSubResource.messageHandler
    at com.spotify.google.cloud.pubsub.client.Puller.lambda$pullBatch$11
    at java.util.concurrent.CompletableFuture.uniWhenComplete
    at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire
    at java.util.concurrent.CompletableFuture.postComplete
    at java.util.concurrent.CompletableFuture.complete
    at com.spotify.google.cloud.pubsub.client.PubsubFuture.succeed
    at com.spotify.google.cloud.pubsub.client.PubsubFuture.lambda$wrap$9
    at java.util.concurrent.CompletableFuture.uniWhenComplete
    at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire
    at java.util.concurrent.CompletableFuture.postComplete
    at java.util.concurrent.CompletableFuture.complete
    at com.spotify.google.cloud.pubsub.client.PubsubFuture.succeed
    at com.spotify.google.cloud.pubsub.client.Pubsub.lambda$requestJavaNet$8
    at java.util.concurrent.ThreadPoolExecutor.runWorker
    at java.util.concurrent.ThreadPoolExecutor$Worker.run
    at java.lang.Thread.run
```

### PR DESCRIPTION
**Pull request by mziccard**
This fixes #1180 and #1018

### CODE REVIEW

**Comment by mziccard** [comment]
FYI @fonzy2013

@aozarov Do you mind having a look? This also retries the 'insufficient data written' exception, more background can be found at #1018

File: gcloud-java-core/src/main/java/com/google/cloud/BaseServiceException.java [hunk 1]
```diff
@@ -187,7 +188,10 @@ protected boolean isRetryable(boolean idempotent, Error error) {
   }
 
   protected boolean isRetryable(boolean idempotent, IOException exception) {
-    return idempotent && exception instanceof SocketTimeoutException;
+    boolean exceptionIsRetryable = exception instanceof SocketTimeoutException
+        || exception instanceof SocketException
+        || "insufficient data written".equals(exception.getMessage());
```
Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/cloud/BaseServiceException.java]: Lets hope that a retry indeed fix the "insufficient data written" case (which sounds like a bug in the google-api-java-client, however I am not sure what actually triggers it).

**Comment by aozarov** [comment]
LGTM

File: gcloud-java-core/src/main/java/com/google/cloud/BaseServiceException.java (hunk 1 above)
Comment (mziccard) [inline gcloud-java-core/src/main/java/com/google/cloud/BaseServiceException.java]: > Lets hope that a retry indeed fix the "insufficient data written" case (which sounds like a bug in the google-api-java-client, however I am not sure what actually triggers it).

That exception does not occur systematically, retrying a request does not trigger the same error, and in fact it usually succeeds. I am still unsure about the cause though.

**Comment by fonzy2013** [comment]
@mziccard which release should I use for this fix? 0.2.8 or 0.29-SNAPSHOT?

**Comment by mziccard** [comment]
@fonzy2013 I wrote you [here](https://github.com/GoogleCloudPlatform/gcloud-java/issues/1180#issuecomment-242732843) on Friday (its seems notification got lost :(): `0.2.8` contains this fix. For more details on the changes contained in `0.2.8` have a look at our [release notes](https://github.com/GoogleCloudPlatform/gcloud-java/releases/tag/v0.2.8)

**Pull request merged**